rfcs/20200712-tfrt-kernel-fallback.md

```diff
@@ -0,0 +1,647 @@
+    Fallback. Adding inheritance would impact all existing TensorFlow kernels
+    even those that don't support Kernel Fallback.
+*   Binary size increase when using templates compared to inheritance is
+    estimated at 2.6% (based on adding `AddN` op).
+
+Right now, we are leaning towards using inheritance. Seems like time increase is
+only significant for running many scalar ops in a sequence - probably a rare use
+case in the real world.
+
+To use inheritance, we will define `OpKernelConstructionInterface` and
+`OpKernelContextInterface` interfaces. Ideally, these interfaces should be pure
+virtual. However, we will have one exception - templated `eigen_device` method
+that calls per-device pure-virtual implementations.
+
+We will then introduce `TFRTOpKernelConstruction` and `TFRTOpKernelContext`
+subclasses that implement `OpKernelConstructionInterface` and
+`OpKernelContextInterface` in terms of TFRT data structures. Example how
+`TFRTOpKernelConstruction` might look like:
+
+```cpp
+class TFRTOpKernelConstruction : public OpKernelConstructionInterface {
+ public:
+  explicit TFRTOpKernelConstruction(AttrMap attributes);
+  ~TFRTOpKernelConstruction() override {};
+
+  Status GetAttr(StringPiece attr_name, int32* value) const override;
+  Status GetAttr(StringPiece attr_name, DataType* value) const override;
+
+  void CtxFailure(const Status& s);
+  void CtxFailureWithWarning(const Status& s);
+  void CtxFailure(const char* file, int line, const Status& s);
+  void CtxFailureWithWarning(const char* file, int line, const Status& s);
+  ...
+};
+```
+
+When forwarding, we instantiate the kernel interfaces with TFRT’s lightweight
+OpKernel definitions, rather than TensorFlow’s
+[heavyweight OpKernel definitions](https://cs.opensource.google/android/platform/superproject/+/master:external/tensorflow/tensorflow/core/framework/op_kernel.h;l=612?q=opkernelcontext)
+for example.
+
+Example `AddN` kernel implementation using these new interfaces:
+
+```cpp
+class AddNOp : public OpKernelBase {
+ public:
+   explicit AddNOp(OpKernelConstructionInterface* construction) :
+       OpKernelBase(construction) {}
+
+   void Compute(OpKernelContextInterface* ctx) override {
+     if(!ctx->ValidateInputsAreSameShape(this)) return;
+     ...
+```
+
+Here, `OpKernelBase` implementation will be minimal:
+
+```cpp
+class OpKernelBase {
+ public:
+  explicit OpKernelBase(OpKernelConstructionInterface* context) {
+  }
+  virtual ~OpKernelBase() {}
+  virtual void Compute(OpKernelContextInterface* context) = 0;
+};
+```
+
+(For details how extending from `OpKernelBase` instead of `OpKernel` would work
+with current TensorFlow runtime see [Appendix 1](#appendix-1))
+
+Corresponding .cc file then registers the kernel using the correct kernel and
+context classes. For example, this is how we register `AddN` kernel with TFRT:
+
+```cpp
+REGISTER_KERNEL_FALLBACK_KERNEL( "AddN", AddNOp<CPUDevice, int32>);
```

Comment:
to confirm, will we be adding these extra symbols to TF, where a CPU kernel like "AddN" for fallback is different from the original CPU kernel for "AddN", or will we be upgrading the existing symbol with fallback capability? (I think it's the latter)

For that CPU kernel, will we still need to use the original REGISTER_KERNEL, or just the new  REGISTER_KERNEL_FALLBACK_KERNEL? (I hope it's the latter)

Reply:
We will be upgrading existing kernel code, but we will need 2 separate registrations:

* `REGISTER_KERNEL_FALLBACK_KERNEL`: register kernel with Kernel Fallback.
* `REGISTER_KERNEL_BASE_BUILDER`: register a kernel with current TensorFlow runtime. Note this is different from `REGISTER_KERNEL_BUILDER` that TensorFlow currently uses. We are using a separate kernel registration factory due to reasons in Appendix 1 (we wrap `OpKernelBase` with `OpKernel`).

Reason for 2 registrations:
Current TensorFlow registration mechanism is considered part of *runtime* and we are avoiding dependency on runtime to reduce code size. I suppose we could still have one macro that calls both registrations but we need to make sure it doesn't call TF registration if *only* using Kernel Fallback.